src/library/scala/util/Using.scala

```diff
@@ -0,0 +1,287 @@
+    * @throws java.lang.IllegalStateException if the resource has already been used
+    * @return a [[scala.util.Try `Try`]] containing the result of the operation, or
+    *         an exception if one was thrown by the operation or by releasing the resource
+    */
+  @throws[IllegalStateException]("if the resource has already been used")
+  def map[A](f: R => A)(implicit r: Using.Resource[R]): Try[A] = Try { useWith(f) }
+
+  /** Performs an operation which returns a [[scala.util.Try `Try`]] using a resource,
+    * and then releases the resource, even if the operation throws an exception.
+    *
+    * @param f the `Try`-returning operation to perform
+    * @param r an implicit [[Using.Resource]]
+    * @tparam A the return type of the operation
+    * @throws java.lang.IllegalStateException if the resource has already been used
+    * @return the result of the inner operation, or a [[scala.util.Try `Try`]]
+    *         containing an exception if one was thrown by the operation or by
+    *         releasing the resource
+    */
+  @throws[IllegalStateException]("if the resource has already been used")
+  def flatMap[A](f: R => Try[A])(implicit r: Using.Resource[R]): Try[A] =
+    map {
+      r => f(r).get // otherwise inner Failure will be lost on exceptional release
+    }
+
+  @inline private[this] def useWith[A](f: R => A)(implicit r: Using.Resource[R]): A =
+    if (used.getAndSet(true)) throw new IllegalStateException("resource has already been used")
+    else Using.resource(resource)(f)
+}
+
+/** @define recommendUsing                   It is highly recommended to use the `Using` construct,
+  *                                          which safely wraps resource usage and management in a `Try`.
+  * @define multiResourceSuppressionBehavior If more than one exception is thrown by the operation and releasing resources,
+  *                                          the exception thrown ''first'' is thrown, with the other exceptions
+  *                                          [[java.lang.Throwable.addSuppressed(Throwable) added as suppressed exceptions]]
+  *                                          to the one thrown first. This is the case ''unless'' a later exception is
+  *                                          [[scala.util.control.NonFatal fatal]], and the one preceding it is not. In that case,
+  *                                          the first exception is added as a suppressed exception to the fatal one, and the fatal
+  *                                          one is thrown. If an exception is a
+  *                                          [[scala.util.control.ControlThrowable ControlThrowable]], no exception will be added to
+  *                                          it as a suppressed exception.
+  */
+object Using {
+  /** Creates a `Using` from the given resource.
+    *
+    * @note If the resource does not have an implicit [[Resource]] in
+    *       scope, the returned `Using` will be useless.
```

Comment:
I think this note might be hard to understand. What does that mean "the resource has an implicit `Resource` in scope"? How will the result be useless?

Reply:
@lrytz In order to allow you to write `Using(foo) { _.bar }`, the implicit `Resource` is a parameter for the `apply`, `map` and `flatMap` methods, rather than the construction of the `Using` (see [this comment from @Ichoran](https://github.com/scala/scala/pull/6907#issuecomment-403634312)). I was trying to get across that this makes it possible to create a `Using` for which there is no `Resource` (e.g. `Using("foo")` will compile), but that you won't be able to actually use it for anything. I'm happy to reword it if you have suggestions.

Reply:
Yes, I think this is the kind of explanation that that (I think) should be in the scaladoc.

Reply:
https://github.com/NthPortal/scala/commit/ef3b8d495fe25bdec5d439b85faba8532075e06c